pkg/cloud-init/cloud-init.go

```diff
@@ -456,124 +468,149 @@ func GenerateLocalData(vmiName string, namespace string, data *CloudInitData) er
-	files = append(files, metaFile)
-	files = append(files, userFile)
+	files = append(files, dataSourceData.MetaFile)
+	files = append(files, dataSourceData.UserFile)
 
 	if len(networkData) > 0 {
-		err = ioutil.WriteFile(networkFile, networkData, 0644)
+		err = ioutil.WriteFile(dataSourceData.NetworkFile, networkData, 0644)
 		if err != nil {
 			return err
 		}
-		files = append(files, networkFile)
+		files = append(files, dataSourceData.NetworkFile)
 	}
 
 	switch data.DataSource {
 	case DataSourceNoCloud:
-		err = cloudInitIsoFunc(isoStaging, "cidata", dataBasePath)
+		err = cloudInitIsoFunc(dataSourceData.IsoStaging, "cidata", dataSourceData.DataBasePath)
 	case DataSourceConfigDrive:
-		err = cloudInitIsoFunc(isoStaging, "config-2", dataBasePath)
+		err = cloudInitIsoFunc(dataSourceData.IsoStaging, "config-2", dataSourceData.DataBasePath)
 	}
 	if err != nil {
 		return err
 	}
-	diskutils.RemoveFile(metaFile)
-	diskutils.RemoveFile(userFile)
-	diskutils.RemoveFile(networkFile)
+	diskutils.RemoveFile(dataSourceData.MetaFile)
+	diskutils.RemoveFile(dataSourceData.UserFile)
+	diskutils.RemoveFile(dataSourceData.NetworkFile)
 
-	if err := diskutils.DefaultOwnershipManager.SetFileOwnership(isoStaging); err != nil {
+	if err := diskutils.DefaultOwnershipManager.SetFileOwnership(dataSourceData.IsoStaging); err != nil {
 		return err
 	}
 
-	err = os.Rename(isoStaging, iso)
+	err = os.Rename(dataSourceData.IsoStaging, dataSourceData.Iso)
 	if err != nil {
-		log.Log.Reason(err).Errorf("Cloud-init failed to rename file %s to %s", isoStaging, iso)
+		log.Log.Reason(err).Errorf("Cloud-init failed to rename file %s to %s", dataSourceData.IsoStaging, dataSourceData.Iso)
 		return err
 	}
 
-	log.Log.V(2).Infof("generated nocloud iso file %s", iso)
+	log.Log.V(2).Infof("generated nocloud iso file %s", dataSourceData.Iso)
 	return nil
 }
 
+func generateLocalDataSourceData(vmiName string, namespace string, data *CloudInitData) (*LocalDataSourceData, error) {
+	var err error
+	dataSourceData := &LocalDataSourceData{}
+	dataSourceData.DomainBasePath = getDomainBasePath(vmiName, namespace)
+	dataSourceData.DataBasePath = fmt.Sprintf("%s/data", dataSourceData.DomainBasePath)
+
+	if data.DataSource == DataSourceNoCloud {
+		dataSourceData, err = generateLocalDataSourceNoCloudData(vmiName, namespace, data, dataSourceData)
+		if err != nil {
+			return nil, err
+		}
+	} else if data.DataSource == DataSourceConfigDrive {
+		dataSourceData, err = generateLocalDataSourceConfigDriveData(vmiName, namespace, data, dataSourceData)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		return nil, fmt.Errorf("Invalid cloud-init data source: '%v'", data.DataSource)
+	}
+	return dataSourceData, nil
+}
+
+func generateLocalDataSourceNoCloudData(vmiName string, namespace string, data *CloudInitData, d *LocalDataSourceData) (*LocalDataSourceData, error) {
+	var err error
+	d.DataPath = d.DataBasePath
+	d.MetaFile = fmt.Sprintf("%s/%s", d.DataPath, "meta-data")
```

Comment:
Better use filepath.Join instead of Sprintf

Reply:
Thanks! [Updated](https://github.com/kubevirt/kubevirt/compare/f8f710cf948b7815404804db18e60e7272bca22e..22bf9ea2acdaf7c5837f85ff759737945737399c) 

I have also changed namimg a bit. Namely removed the trailing `Data`, because it's obvious it contains data.

Reply:
The `GenerateLocalData`  is generating real files on the local FS while your new functions simply create data structures, that are then used to generate the local files. 
Maybe change the structure name to just `DataSource`  (or `CloudInitDataSource`) and then change your functions to` createNoCloudDataCloudInitDataSource`  and `createConfigDriveCloudInitDataSource` .